internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -677,12 +702,41 @@ func (r *ClusterDeploymentsReconciler) deregisterClusterIfNeeded(ctx context.Con
 	}); err != nil {
 		return buildReply(err)
 	}
+	// Delete agents because their backend cluster got deregistered.
+	if err = r.DeleteClusterDeploymentAgents(ctx, key); err != nil {
+		return buildReply(err)
+	}
 
 	r.Log.Infof("Cluster resource deleted, Unregistered cluster: %s", c.ID.String())
 
 	return buildReply(nil)
 }
 
+func (r *ClusterDeploymentsReconciler) DeleteClusterDeploymentAgents(ctx context.Context, clusterDeployment types.NamespacedName) error {
+	agents := &aiv1beta1.AgentList{}
+	if err := r.List(context.Background(), agents); err != nil {
+		return err
+	}
+	for _, clusterAgent := range agents.Items {
+		if clusterAgent.Spec.ClusterDeploymentName.Name == clusterDeployment.Name &&
+			clusterAgent.Spec.ClusterDeploymentName.Namespace == clusterDeployment.Namespace {
+			r.Log.Infof("clusterDeployment %s namespace %s, delete agent %s",
```

Comment:
The agent namespace is also useful. Since this is a lot of info, maybe put the identifiers for the resources into fields on the logger. Or you could at least do that at the top of this function for the CD, and then identify the Agent in the message if you want.

Reply:
Done